### PR TITLE
bpo-32949: Simplify "with"-related opcodes.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue32949

### CODE REVIEW

**Review by rhettinger — APPROVED** [review]

File: Objects/frameobject.c [hunk 1]
```diff
@@ -232,7 +232,7 @@ frame_setlineno(PyFrameObject *f, PyObject* p_new_lineno)
                     delta_iblock++;
                 }
             }
-            if (op != FOR_ITER) {
+            if (op == SETUP_FINALLY) {
```
Comment (pitrou) [inline Objects/frameobject.c]: Is this change related to this PR or was it a lingering bug?

File: Python/ceval.c [hunk 2]
```diff
@@ -2955,97 +2954,57 @@ _PyEval_EvalFrameDefault(PyFrameObject *f, int throwflag)
...
-            stack[2] = tb;
-            res = _PyObject_FastCall(exit_func, stack, 3);
+               We shift the top 6 values of the stack down and push the result
+               of the call EXIT(TOP, SECOND, THIRD).
+            */
+            PyObject *exc = TOP();
+            assert(PyExceptionClass_Check(exc));
+            PyObject *val = SECOND();
+            PyObject *tb = THIRD();
+            PyObject *tp2 = FOURTH();
```
Comment (pitrou) [inline Python/ceval.c]: There's `exc, val, tb` above so I would expect `exc2, val2, tb2` below.

File: Python/ceval.c [hunk 3]
```diff
@@ -2955,97 +2954,57 @@ _PyEval_EvalFrameDefault(PyFrameObject *f, int throwflag)
...
+            PyObject *exc2 = PEEK(5);
+            PyObject *tb2 = PEEK(6);
+            PyObject *exit_func = PEEK(7);
+            SET_VALUE(7, tb2);
+            SET_VALUE(6, exc2);
+            SET_VALUE(5, tp2);
+            SET_FOURTH(tb);
+            SET_THIRD(val);
+            SET_SECOND(exc);
+            /* We just shifted the stack down, so we have
```
Comment (pitrou) [inline Python/ceval.c]: Hmm... instead of fixing the except handler block, can't we put a NULL or None at the seventh stack slot? That would feel less hackish.

File: Python/compile.c [hunk 4]
```diff
@@ -4474,10 +4489,17 @@ compiler_with(struct compiler *c, stmt_ty s, int pos)
...
+    ADDOP_N(c, LOAD_CONST, args, consts);
     ADDOP(c, POP_BLOCK);
-    ADDOP(c, BEGIN_FINALLY);
     compiler_pop_fblock(c, WITH, block);
+    ADDOP_I(c, CALL_FUNCTION_EX, 0);
+    ADDOP(c, POP_TOP);
+    ADDOP_JREL(c, JUMP_FORWARD, end);
 
     compiler_use_next_block(c, finally);
     if (!compiler_push_fblock(c, FINALLY_END, finally, NULL))
```
Comment (pitrou) [inline Python/compile.c]: Can you update the comment three lines below this? Currently it says:
```
     /* Finally block starts; context.__exit__ is on the stack under
       the exception or return information.
````

I would expect "Except block" instead of "Finally block", also "return information" doesn't apply anymore.

Comment (pitrou) [reply Python/compile.c]: Same comment for `compiler_async_with`, btw.

File: Objects/frameobject.c (hunk 1 above)
Comment (serhiy-storchaka) [reply Objects/frameobject.c]: Yes, it is related. Prior this change SETUP_WITH and SETUP_ASYNC_WITH blocks were followed by finally block ended by END_FINALLY. Even if this block was atomic (because it didn't correspond any line of code) we pushed its start into blockstack for compensating  END_FINALLY. Now END_FINALLY will be not used.

File: Python/ceval.c (hunk 2 above)
Comment (serhiy-storchaka) [reply Python/ceval.c]: Agreed. `exc, val, tb` and `tp2, exc2, tb2` looked confusing.

File: Python/ceval.c (hunk 3 above)
Comment (serhiy-storchaka) [reply Python/ceval.c]: Interesting idea. I'll try to implement it.

While I thought about it I have found a bug in existing code (see bpo-33026).

Comment (serhiy-storchaka) [reply Python/ceval.c]: Done.
